src/solvers/solver_utils.hpp

```diff
@@ -146,34 +148,38 @@ struct Stencil {
 };
 
 namespace utils {
-template <class in, class out, bool only_fine_on_composite = true>
+template <class in_t, class out_t, bool only_fine_on_composite = true>
 TaskStatus CopyData(const std::shared_ptr<MeshData<Real>> &md) {
   using TE = parthenon::TopologicalElement;
   TE te = TE::CC;
   IndexRange ib = md->GetBoundsI(IndexDomain::entire, te);
   IndexRange jb = md->GetBoundsJ(IndexDomain::entire, te);
   IndexRange kb = md->GetBoundsK(IndexDomain::entire, te);
 
-  auto desc = parthenon::MakePackDescriptor<in, out>(md.get());
-  auto pack = desc.GetPack(md.get(), {}, only_fine_on_composite);
-  parthenon::par_for(
-      DEFAULT_LOOP_PATTERN, "CopyData", DevExecSpace(), 0, pack.GetNBlocks() - 1, kb.s,
-      kb.e, jb.s, jb.e, ib.s, ib.e,
-      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
-        // TODO(LFR): If this becomes a bottleneck, exploit hierarchical parallelism and
-        //            pull the loop over vars outside of the innermost loop to promote
-        //            vectorization.
-        const int nvars = pack.GetUpperBound(b, in()) - pack.GetLowerBound(b, in()) + 1;
-        for (int c = 0; c < nvars; ++c)
-          pack(b, te, out(c), k, j, i) = pack(b, te, in(c), k, j, i);
+  static auto desc = parthenon::MakePackDescriptor<in_t, out_t>(md.get());
+  auto pack = desc.GetPack(md.get(), only_fine_on_composite);
+  const int scratch_size = 0;
+  const int scratch_level = 0;
+  parthenon::par_for_outer(
```

Comment:
Do you suspect that here a `par_for_outer` is more performant than a flat `par_for` with a standard `for` loop over `nvars` in the kernel?

Reply:
At least on CPU, the inner loop should vectorize. On GPU, I think this should minimize the number of kernel launches. Also, it minimizes the number of times we call the type based accessor, which requires more indirection than directly accessing the pointer.